{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#This was the fix for dca:\n",
    "\n",
    "mamba create -n enimpute python==3.8\n",
    "\n",
    "conda activate enimpute \n",
    "\n",
    "/home/subudhak/miniconda3/envs/enimpute/bin/pip3 install dca 'keras>=2.4,<2.6' 'tensorflow>=2.0,<2.5'\n",
    "\n",
    "https://github.com/theislab/dca/issues/66\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "# from magic import MAGIC\n",
    "import scanpy.external as sce\n",
    "\n",
    "\n",
    "adata = sc.read_h5ad(\"/home/sah2p/ondemand/singlecell_data/SRP173393.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean expression value: 0.05\n"
     ]
    }
   ],
   "source": [
    "adata\n",
    "import numpy as np\n",
    "\n",
    "# Check if the data is normalized\n",
    "mean_expression = np.mean(adata.X)\n",
    "print(f\"Mean expression value: {mean_expression:.2f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using raw count data from adata.raw.X\n"
     ]
    }
   ],
   "source": [
    "if adata.raw is not None:\n",
    "    print(\"Using raw count data from adata.raw.X\")\n",
    "adata.X = adata.raw.X.copy()\n",
    "    \n",
    "# Remove all-zero genes\n",
    "sc.pp.filter_genes(adata, min_counts=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "\n",
    "# Enable TensorFlow 2.x behavior and disable session-based execution\n",
    "tf.compat.v1.disable_eager_execution()\n",
    "\n",
    "# Reset Keras session to avoid variable conflicts\n",
    "tf.keras.backend.clear_session()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dca: Successfully preprocessed 34017 genes and 13252 cells.\n",
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-04 10:19:48.152264: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"model_3\"\n",
      "__________________________________________________________________________________________________\n",
      "Layer (type)                    Output Shape         Param #     Connected to                     \n",
      "==================================================================================================\n",
      "count (InputLayer)              [(None, 34017)]      0                                            \n",
      "__________________________________________________________________________________________________\n",
      "enc0 (Dense)                    (None, 64)           2177152     count[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization (BatchNorma (None, 64)           192         enc0[0][0]                       \n",
      "__________________________________________________________________________________________________\n",
      "enc0_act (Activation)           (None, 64)           0           batch_normalization[0][0]        \n",
      "__________________________________________________________________________________________________\n",
      "center (Dense)                  (None, 32)           2080        enc0_act[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_1 (BatchNor (None, 32)           96          center[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "center_act (Activation)         (None, 32)           0           batch_normalization_1[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "dec1 (Dense)                    (None, 64)           2112        center_act[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_2 (BatchNor (None, 64)           192         dec1[0][0]                       \n",
      "__________________________________________________________________________________________________\n",
      "dec1_act (Activation)           (None, 64)           0           batch_normalization_2[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "mean (Dense)                    (None, 34017)        2211105     dec1_act[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "size_factors (InputLayer)       [(None, 1)]          0                                            \n",
      "__________________________________________________________________________________________________\n",
      "lambda_1 (Lambda)               (None, 34017)        0           mean[0][0]                       \n",
      "                                                                 size_factors[0][0]               \n",
      "__________________________________________________________________________________________________\n",
      "dispersion (Dense)              (None, 34017)        2211105     dec1_act[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "slice (SliceLayer)              (None, 34017)        0           lambda_1[1][0]                   \n",
      "                                                                 dispersion[0][0]                 \n",
      "==================================================================================================\n",
      "Total params: 6,604,034\n",
      "Trainable params: 6,603,714\n",
      "Non-trainable params: 320\n",
      "__________________________________________________________________________________________________\n",
      "Train on 11926 samples, validate on 1326 samples\n",
      "Epoch 1/10\n",
      "11904/11926 [============================>.] - ETA: 0s - loss: 0.2834"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sah2p/ondemand/tools/miniconda3/envs/dcaimpute/lib/python3.8/site-packages/tensorflow/python/keras/engine/training.py:2323: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.\n",
      "  warnings.warn('`Model.state_updates` will be removed in a future version. '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11926/11926 [==============================] - 11s 947us/sample - loss: 0.2832 - val_loss: 0.1754\n",
      "Epoch 2/10\n",
      "11926/11926 [==============================] - 11s 925us/sample - loss: 0.2199 - val_loss: 0.1694\n",
      "Epoch 3/10\n",
      "11926/11926 [==============================] - 11s 927us/sample - loss: 0.2171 - val_loss: 0.1667\n",
      "Epoch 4/10\n",
      "11926/11926 [==============================] - 11s 935us/sample - loss: 0.2158 - val_loss: 0.1653\n",
      "Epoch 5/10\n",
      "11926/11926 [==============================] - 11s 922us/sample - loss: 0.2148 - val_loss: 0.1642\n",
      "Epoch 6/10\n",
      "11926/11926 [==============================] - 11s 914us/sample - loss: 0.2142 - val_loss: 0.1636\n",
      "Epoch 7/10\n",
      "11926/11926 [==============================] - 11s 902us/sample - loss: 0.2138 - val_loss: 0.1632\n",
      "Epoch 8/10\n",
      "11926/11926 [==============================] - 11s 897us/sample - loss: 0.2134 - val_loss: 0.1631\n",
      "Epoch 9/10\n",
      "11926/11926 [==============================] - 11s 904us/sample - loss: 0.2131 - val_loss: 0.1629\n",
      "Epoch 10/10\n",
      "11926/11926 [==============================] - 11s 913us/sample - loss: 0.2128 - val_loss: 0.1621\n",
      "dca: Calculating reconstructions...\n"
     ]
    }
   ],
   "source": [
    "sce.pp.dca(\n",
    "    adata,\n",
    "    copy=True,          # Whether to return a copy instead of modifying in place\n",
    "    hidden_size=[64, 32, 64],  # Neural network architecture\n",
    "    epochs=10,         # Number of training epochs\n",
    "    batch_size=32,      # Batch size for training\n",
    "    learning_rate=0.001, # Learning rate for optimization\n",
    "    verbose=True        # Print progress messages\n",
    ")\n",
    "# sce.pp.dca(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>Percent.mt</th>\n",
       "      <th>Seurat_clusters</th>\n",
       "      <th>Celltype</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Libraries</th>\n",
       "      <th>ACE</th>\n",
       "      <th>dca_split</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>size_factors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRX5128615@@_AAACCTGAGAGTAAGG-1</th>\n",
       "      <td>SRX5128615</td>\n",
       "      <td>1131.0</td>\n",
       "      <td>897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Root epidermis</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>1131.0</td>\n",
       "      <td>0.311656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX5128615@@_AAACCTGAGATCCGAG-1</th>\n",
       "      <td>SRX5128615</td>\n",
       "      <td>4346.0</td>\n",
       "      <td>1958</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>23</td>\n",
       "      <td>Root endodermis</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>4346.0</td>\n",
       "      <td>1.197575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX5128615@@_AAACCTGAGCGCCTCA-1</th>\n",
       "      <td>SRX5128615</td>\n",
       "      <td>6849.0</td>\n",
       "      <td>2614</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>16</td>\n",
       "      <td>G2/M phase</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>6849.0</td>\n",
       "      <td>1.887297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX5128615@@_AAACCTGAGGAATGGA-1</th>\n",
       "      <td>SRX5128615</td>\n",
       "      <td>2470.0</td>\n",
       "      <td>1524</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13</td>\n",
       "      <td>Root epidermis</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>2470.0</td>\n",
       "      <td>0.680628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX5128615@@_AAACCTGCACTTAACG-1</th>\n",
       "      <td>SRX5128615</td>\n",
       "      <td>12600.0</td>\n",
       "      <td>3711</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>12</td>\n",
       "      <td>Non-hair</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>12600.0</td>\n",
       "      <td>3.472031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX5128617@@_TTTGGTTGTTACGCGC-1</th>\n",
       "      <td>SRX5128617</td>\n",
       "      <td>1860.0</td>\n",
       "      <td>1041</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>Root cap</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0 (shr-3)</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>1860.0</td>\n",
       "      <td>0.512538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX5128617@@_TTTGTCAGTATGAATG-1</th>\n",
       "      <td>SRX5128617</td>\n",
       "      <td>3580.0</td>\n",
       "      <td>1290</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>22</td>\n",
       "      <td>Unknow</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0 (shr-3)</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>3580.0</td>\n",
       "      <td>0.986498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX5128617@@_TTTGTCAGTGCGATAG-1</th>\n",
       "      <td>SRX5128617</td>\n",
       "      <td>12396.0</td>\n",
       "      <td>2438</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>Lateral root cap</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0 (shr-3)</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>12396.0</td>\n",
       "      <td>3.415817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX5128617@@_TTTGTCATCACGCGGT-1</th>\n",
       "      <td>SRX5128617</td>\n",
       "      <td>3194.0</td>\n",
       "      <td>1329</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>22</td>\n",
       "      <td>Unknow</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0 (shr-3)</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>3194.0</td>\n",
       "      <td>0.880132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX5128617@@_TTTGTCATCATACGGT-1</th>\n",
       "      <td>SRX5128617</td>\n",
       "      <td>1255.0</td>\n",
       "      <td>890</td>\n",
       "      <td>0.159363</td>\n",
       "      <td>13</td>\n",
       "      <td>Root epidermis</td>\n",
       "      <td>SRP173393</td>\n",
       "      <td>Root tip</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0 (shr-3)</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>6 days old seedling</td>\n",
       "      <td>train</td>\n",
       "      <td>1255.0</td>\n",
       "      <td>0.345825</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13252 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "SRX5128615@@_AAACCTGAGAGTAAGG-1  SRX5128615      1131.0           897   \n",
       "SRX5128615@@_AAACCTGAGATCCGAG-1  SRX5128615      4346.0          1958   \n",
       "SRX5128615@@_AAACCTGAGCGCCTCA-1  SRX5128615      6849.0          2614   \n",
       "SRX5128615@@_AAACCTGAGGAATGGA-1  SRX5128615      2470.0          1524   \n",
       "SRX5128615@@_AAACCTGCACTTAACG-1  SRX5128615     12600.0          3711   \n",
       "...                                     ...         ...           ...   \n",
       "SRX5128617@@_TTTGGTTGTTACGCGC-1  SRX5128617      1860.0          1041   \n",
       "SRX5128617@@_TTTGTCAGTATGAATG-1  SRX5128617      3580.0          1290   \n",
       "SRX5128617@@_TTTGTCAGTGCGATAG-1  SRX5128617     12396.0          2438   \n",
       "SRX5128617@@_TTTGTCATCACGCGGT-1  SRX5128617      3194.0          1329   \n",
       "SRX5128617@@_TTTGTCATCATACGGT-1  SRX5128617      1255.0           890   \n",
       "\n",
       "                                 Percent.mt Seurat_clusters          Celltype  \\\n",
       "SRX5128615@@_AAACCTGAGAGTAAGG-1    0.000000               0    Root epidermis   \n",
       "SRX5128615@@_AAACCTGAGATCCGAG-1    0.000000              23   Root endodermis   \n",
       "SRX5128615@@_AAACCTGAGCGCCTCA-1    0.000000              16        G2/M phase   \n",
       "SRX5128615@@_AAACCTGAGGAATGGA-1    0.000000              13    Root epidermis   \n",
       "SRX5128615@@_AAACCTGCACTTAACG-1    0.007937              12          Non-hair   \n",
       "...                                     ...             ...               ...   \n",
       "SRX5128617@@_TTTGGTTGTTACGCGC-1    0.000000               1          Root cap   \n",
       "SRX5128617@@_TTTGTCAGTATGAATG-1    0.000000              22            Unknow   \n",
       "SRX5128617@@_TTTGTCAGTGCGATAG-1    0.000000               2  Lateral root cap   \n",
       "SRX5128617@@_TTTGTCATCACGCGGT-1    0.000000              22            Unknow   \n",
       "SRX5128617@@_TTTGTCATCATACGGT-1    0.159363              13    Root epidermis   \n",
       "\n",
       "                                   Dataset    Tissue Organ Condition  \\\n",
       "SRX5128615@@_AAACCTGAGAGTAAGG-1  SRP173393  Root tip  Root    Normal   \n",
       "SRX5128615@@_AAACCTGAGATCCGAG-1  SRP173393  Root tip  Root    Normal   \n",
       "SRX5128615@@_AAACCTGAGCGCCTCA-1  SRP173393  Root tip  Root    Normal   \n",
       "SRX5128615@@_AAACCTGAGGAATGGA-1  SRP173393  Root tip  Root    Normal   \n",
       "SRX5128615@@_AAACCTGCACTTAACG-1  SRP173393  Root tip  Root    Normal   \n",
       "...                                    ...       ...   ...       ...   \n",
       "SRX5128617@@_TTTGGTTGTTACGCGC-1  SRP173393  Root tip  Root    Normal   \n",
       "SRX5128617@@_TTTGTCAGTATGAATG-1  SRP173393  Root tip  Root    Normal   \n",
       "SRX5128617@@_TTTGTCAGTGCGATAG-1  SRP173393  Root tip  Root    Normal   \n",
       "SRX5128617@@_TTTGTCATCACGCGGT-1  SRP173393  Root tip  Root    Normal   \n",
       "SRX5128617@@_TTTGTCATCATACGGT-1  SRP173393  Root tip  Root    Normal   \n",
       "\n",
       "                                      Genotype     Libraries  \\\n",
       "SRX5128615@@_AAACCTGAGAGTAAGG-1          Col-0  10x Genomics   \n",
       "SRX5128615@@_AAACCTGAGATCCGAG-1          Col-0  10x Genomics   \n",
       "SRX5128615@@_AAACCTGAGCGCCTCA-1          Col-0  10x Genomics   \n",
       "SRX5128615@@_AAACCTGAGGAATGGA-1          Col-0  10x Genomics   \n",
       "SRX5128615@@_AAACCTGCACTTAACG-1          Col-0  10x Genomics   \n",
       "...                                        ...           ...   \n",
       "SRX5128617@@_TTTGGTTGTTACGCGC-1  Col-0 (shr-3)  10x Genomics   \n",
       "SRX5128617@@_TTTGTCAGTATGAATG-1  Col-0 (shr-3)  10x Genomics   \n",
       "SRX5128617@@_TTTGTCAGTGCGATAG-1  Col-0 (shr-3)  10x Genomics   \n",
       "SRX5128617@@_TTTGTCATCACGCGGT-1  Col-0 (shr-3)  10x Genomics   \n",
       "SRX5128617@@_TTTGTCATCATACGGT-1  Col-0 (shr-3)  10x Genomics   \n",
       "\n",
       "                                                 ACE dca_split  n_counts  \\\n",
       "SRX5128615@@_AAACCTGAGAGTAAGG-1  6 days old seedling     train    1131.0   \n",
       "SRX5128615@@_AAACCTGAGATCCGAG-1  6 days old seedling     train    4346.0   \n",
       "SRX5128615@@_AAACCTGAGCGCCTCA-1  6 days old seedling     train    6849.0   \n",
       "SRX5128615@@_AAACCTGAGGAATGGA-1  6 days old seedling     train    2470.0   \n",
       "SRX5128615@@_AAACCTGCACTTAACG-1  6 days old seedling     train   12600.0   \n",
       "...                                              ...       ...       ...   \n",
       "SRX5128617@@_TTTGGTTGTTACGCGC-1  6 days old seedling     train    1860.0   \n",
       "SRX5128617@@_TTTGTCAGTATGAATG-1  6 days old seedling     train    3580.0   \n",
       "SRX5128617@@_TTTGTCAGTGCGATAG-1  6 days old seedling     train   12396.0   \n",
       "SRX5128617@@_TTTGTCATCACGCGGT-1  6 days old seedling     train    3194.0   \n",
       "SRX5128617@@_TTTGTCATCATACGGT-1  6 days old seedling     train    1255.0   \n",
       "\n",
       "                                 size_factors  \n",
       "SRX5128615@@_AAACCTGAGAGTAAGG-1      0.311656  \n",
       "SRX5128615@@_AAACCTGAGATCCGAG-1      1.197575  \n",
       "SRX5128615@@_AAACCTGAGCGCCTCA-1      1.887297  \n",
       "SRX5128615@@_AAACCTGAGGAATGGA-1      0.680628  \n",
       "SRX5128615@@_AAACCTGCACTTAACG-1      3.472031  \n",
       "...                                       ...  \n",
       "SRX5128617@@_TTTGGTTGTTACGCGC-1      0.512538  \n",
       "SRX5128617@@_TTTGTCAGTATGAATG-1      0.986498  \n",
       "SRX5128617@@_TTTGTCAGTGCGATAG-1      3.415817  \n",
       "SRX5128617@@_TTTGTCATCACGCGGT-1      0.880132  \n",
       "SRX5128617@@_TTTGTCATCATACGGT-1      0.345825  \n",
       "\n",
       "[13252 rows x 16 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var.index.name = 'genes'\n",
    "adata.obs.index.name = 'cells'\n",
    "adata.raw.var.index.name = 'genes'\n",
    "\n",
    "# adata.var.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "\n",
    "# Enable TensorFlow 2.x behavior and disable session-based execution\n",
    "tf.compat.v1.disable_eager_execution()\n",
    "\n",
    "# Reset Keras session to avoid variable conflicts\n",
    "tf.keras.backend.clear_session()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping `_index` from adata.raw.var\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "can't set attribute",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[57], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m adata\u001b[38;5;241m.\u001b[39mraw \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_index\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m adata\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39mcolumns:\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDropping `_index` from adata.raw.var\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 3\u001b[0m     \u001b[43madata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mraw\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar\u001b[49m \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39mdrop(columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_index\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "\u001b[0;31mAttributeError\u001b[0m: can't set attribute"
     ]
    }
   ],
   "source": [
    "if adata.raw is not None and '_index' in adata.raw.var.columns:\n",
    "    print(\"Dropping `_index` from adata.raw.var\")\n",
    "    adata.raw.var = adata.raw.var.drop(columns=['_index'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking problematic column in adata.var: False\n",
      "Checking problematic column in adata.raw.var: True\n",
      "adata.var columns: Index(['features'], dtype='object')\n",
      "adata.raw.var columns: Index(['_index'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(\"Checking problematic column in adata.var:\", \"_index\" in adata.var.columns)\n",
    "print(\"Checking problematic column in adata.raw.var:\", \"_index\" in adata.raw.var.columns if adata.raw else \"adata.raw is None\")\n",
    "print(\"adata.var columns:\", adata.var.columns)\n",
    "print(\"adata.raw.var columns:\", adata.raw.var.columns if adata.raw else \"No raw.var\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw.var.columns = ['genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# magic_op = MAGIC()\n",
    "# adata.layers[\"MAGIC\"] = magic_op.fit_transform(adata.X)\n",
    "\n",
    "# Save the MAGIC-imputed dataset\n",
    "adata.write_h5ad(\"/home/sah2p/ondemand/singlecell_data/SRP173393magic_d.h5ad\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dcaimpute",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
