{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import argparse\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "import tracemalloc  # For memory profiling\n",
    "from scipy import sparse\n",
    "import scanpy.external as sce\n",
    "import numpy as np\n",
    "import doubletdetection\n",
    "import warnings\n",
    "import logging\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n",
    "from numba.core.errors import NumbaDeprecationWarning\n",
    "warnings.simplefilter(\"ignore\", category=NumbaDeprecationWarning)\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(\n",
    "    format='%(asctime)s - %(levelname)s - %(message)s',\n",
    "    level=logging.INFO\n",
    ")\n",
    "\n",
    "sc.settings.verbosity = 0\n",
    "sc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n",
    "\n",
    "log = logging.getLogger(\"clustering.log\")\n",
    "log.setLevel(logging.INFO)\n",
    "\n",
    "context_path =\"/home/sah2p/ondemand/singlecell_data/Arabidopsis/\"\n",
    "dataset = \"SRP166333\"\n",
    "filename = dataset+\"_bbknn.h5ad\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(context_path+filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n"
     ]
    }
   ],
   "source": [
    "# Using the igraph implementation and a fixed number of iterations can be significantly faster, especially for larger datasets\n",
    "sc.tl.leiden(adata, n_iterations=2)\n",
    "sc.tl.louvain(adata)\n",
    "sc.pl.umap(adata, color=[\"leiden\"], save=dataset+\"_bbknn_leiden.png\")\n",
    "sc.pl.umap(adata, color=[\"louvain\"], save=dataset+\"_bbknn_louvain.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n"
     ]
    }
   ],
   "source": [
    "# Re assessing the doublet detection\n",
    "\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\"leiden\", \"doublet\", \"doublet_score\"],\n",
    "    # increase horizontal space between panels\n",
    "    save = dataset+\"_leiden_bbknn_doublet.png\",\n",
    "    wspace=0.5,\n",
    "    size=3,\n",
    ")\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\"leiden\", \"log1p_total_counts\", \"pct_counts_mt\", \"log1p_n_genes_by_counts\"],\n",
    "    wspace=0.5,\n",
    "    save = dataset+\"_leiden_bbknn_qc.png\",\n",
    "    ncols=2,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Doing Clustering on differnt resolutions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, key_added=\"leiden_res0_1\", resolution=0.1)\n",
    "sc.tl.leiden(adata, key_added=\"leiden_res0_5\", resolution=0.5)\n",
    "sc.tl.leiden(adata, key_added=\"leiden_res1\", resolution=1)\n",
    "\n",
    "sc.tl.louvain(adata, key_added=\"louvain_res0_1\", resolution=0.1)\n",
    "sc.tl.louvain(adata, key_added=\"louvain_res0_5\", resolution=0.5)\n",
    "sc.tl.louvain(adata, key_added=\"louvain_res1\", resolution=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[<AxesSubplot: title={'center': 'leiden_res0_1'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'leiden_res0_5'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'leiden_res1'}, xlabel='UMAP1', ylabel='UMAP2'>]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\"leiden_res0_1\", \"leiden_res0_5\",\"leiden_res1\"],\n",
    "    legend_loc=\"on data\",\n",
    "    show = False,\n",
    "    save = dataset+\"_bbknn_leiden_resolutions.png\",)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[<AxesSubplot: title={'center': 'louvain_res0_1'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'louvain_res0_5'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'louvain_res1'}, xlabel='UMAP1', ylabel='UMAP2'>]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\"louvain_res0_1\", \"louvain_res0_5\", \"louvain_res1\"],\n",
    "    legend_loc=\"on data\",\n",
    "    show = False,\n",
    "    save = dataset+\"_bbknn_louvain_resolutions.png\",)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/numpy/core/fromnumeric.py:84: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    }
   ],
   "source": [
    "# Perform differential expression analysis\n",
    "cluster_column = \"leiden_res0_5\"\n",
    "sc.tl.rank_genes_groups(adata, groupby=cluster_column, method=\"wilcoxon\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting ace_tools\n",
      "  Downloading ace_tools-0.0-py3-none-any.whl.metadata (300 bytes)\n",
      "Downloading ace_tools-0.0-py3-none-any.whl (1.1 kB)\n",
      "Installing collected packages: ace_tools\n",
      "Successfully installed ace_tools-0.0\n"
     ]
    }
   ],
   "source": [
    "!pip install ace_tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top markers saved to /home/sah2p/ondemand/singlecell_data/Arabidopsis/top_markers_per_cluster_scanpy.csv\n"
     ]
    }
   ],
   "source": [
    "# Extract the results\n",
    "top_n = 10  # Number of top markers per cluster\n",
    "ranked_genes = pd.DataFrame()\n",
    "\n",
    "for cluster in adata.obs[cluster_column].unique():\n",
    "    top_genes = pd.DataFrame({\n",
    "        \"gene\": adata.uns[\"rank_genes_groups\"][\"names\"][cluster][:top_n],\n",
    "        \"log2FC\": adata.uns[\"rank_genes_groups\"][\"logfoldchanges\"][cluster][:top_n],\n",
    "        \"p_val\": adata.uns[\"rank_genes_groups\"][\"pvals\"][cluster][:top_n],\n",
    "        \"p_val_adj\": adata.uns[\"rank_genes_groups\"][\"pvals_adj\"][cluster][:top_n],\n",
    "        \"cluster\": cluster\n",
    "    })\n",
    "    ranked_genes = pd.concat([ranked_genes, top_genes])\n",
    "\n",
    "# Save to CSV\n",
    "output_file = context_path+\"top_markers_per_cluster_scanpy.csv\"\n",
    "ranked_genes.to_csv(output_file, index=False)\n",
    "\n",
    "print(f\"Top markers saved to {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_dotplot.py:168: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  obs_bool.groupby(level=0).sum() / obs_bool.groupby(level=0).count()\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_dotplot.py:178: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  dot_color_df = self.obs_tidy.groupby(level=0).mean()\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_dotplot.py:747: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap', 'norm' will be ignored\n",
      "  dot_ax.scatter(x, y, **kwds)\n"
     ]
    }
   ],
   "source": [
    "sc.pl.dotplot(adata, var_names=ranked_genes[\"gene\"].unique(), groupby=cluster_column, save=dataset+\"_clusterGenes_dotplot.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering out marker genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(context_path+\"arabidopsis_thaliana.marker_fd.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(141696, 13)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out weak markers:\n",
    "# Remove genes with low avg_log2FC (e.g., below 0.5).\n",
    "# Remove genes with low p_val_adj (e.g., above 0.05).\n",
    "df = df[(df[\"avg_log2FC\"] > 0.5) & (df[\"p_val_adj\"] < 0.05)]\n",
    "df.shape\n",
    "# Filter out genes that are not in the adata object:\n",
    "# df = df[df[\"gene\"].isin(adata.var_names)]\n",
    "#Filter my dataset\n",
    "df = df[df['dataset'] == 'SRP166333']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "TOP_N_GENES = 10\n",
    "# Get the top N genes for each cluster\n",
    "top_genes = df.groupby(\"clusterName\").apply(lambda x: x.nlargest(TOP_N_GENES, \"avg_log2FC\"))\n",
    "top_genes = top_genes.reset_index(drop=True)\n",
    "# Save the ranked genes per cluster to a CSV file\n",
    "output_file = context_path+\"top_markers_per_cluster.csv\"\n",
    "top_genes.to_csv(output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset to only the markers that are in the data\n",
    "marker_genes_in_data = {}\n",
    "for ct, markers in marker_genes.items():\n",
    "    markers_found = []\n",
    "    for marker in markers:\n",
    "        if marker in adata.var.index:\n",
    "            markers_found.append(marker)\n",
    "    marker_genes_in_data[ct] = markers_found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from itertools import combinations\n",
    "from scipy.spatial.distance import jaccard\n",
    "\n",
    "# Load clustered genes (Your dataset)\n",
    "clustered_gene_file = \"/home/sah2p/ondemand/singlecell_data/Arabidopsis/top_markers_per_cluster_scanpy.csv\"  # Replace with your file\n",
    "cluster_df = pd.read_csv(clustered_gene_file)  # Assumes two columns: \"cluster\" & \"gene\"\n",
    "\n",
    "# Load top marker genes per cell type (Preprocessed file from previous steps)\n",
    "marker_gene_file = \"/home/sah2p/ondemand/singlecell_data/Arabidopsis/top_markers_per_cluster.csv\"  # Replace with your file\n",
    "marker_df = pd.read_csv(marker_gene_file)  # Assumes two columns: \"celltype_id\" & \"gene\"\n",
    "\n",
    "# # Convert to dictionary for easy lookup\n",
    "cluster_dict = cluster_df.groupby(\"cluster\")[\"gene\"].apply(set).to_dict()\n",
    "celltype_dict = marker_df.groupby(\"clusterName\")[\"gene\"].apply(set).to_dict()\n",
    "\n",
    "# Compute Jaccard Similarity\n",
    "cluster_annotations = {}\n",
    "\n",
    "for cluster, cluster_genes in cluster_dict.items():\n",
    "    best_match = None\n",
    "    best_score = 0\n",
    "    \n",
    "    for celltype, celltype_genes in celltype_dict.items():\n",
    "        intersection = len(cluster_genes & celltype_genes)\n",
    "        union = len(cluster_genes | celltype_genes)\n",
    "        score = intersection / union if union > 0 else 0  # Jaccard similarity\n",
    "\n",
    "        if score > best_score:\n",
    "            best_score = score\n",
    "            best_match = celltype\n",
    "\n",
    "    cluster_annotations[cluster] = {\"best_match\": best_match, \"jaccard_score\": best_score}\n",
    "\n",
    "# # Convert results to DataFrame\n",
    "annotations_df = pd.DataFrame.from_dict(cluster_annotations, orient=\"index\")\n",
    "annotations_df.reset_index(inplace=True)\n",
    "annotations_df.columns = [\"cluster\", \"predicted_celltype\", \"jaccard_score\"]\n",
    "\n",
    "# # Save to CSV\n",
    "# annotations_df.to_csv(\"cluster_annotations.csv\", index=False)\n",
    "\n",
    "# # Display results\n",
    "# import ace_tools as tools\n",
    "# tools.display_dataframe_to_user(name=\"Cluster Annotations\", dataframe=annotations_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations_df.to_csv(context_path+\"cluster_annotations.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>p_val</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT2G46890</td>\n",
       "      <td>5.562217</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT3G48185</td>\n",
       "      <td>5.308359</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT2G40480</td>\n",
       "      <td>4.601549</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G25530</td>\n",
       "      <td>5.839280</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT5G57770</td>\n",
       "      <td>5.368637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>AT5G22440</td>\n",
       "      <td>3.521461</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>AT3G58700</td>\n",
       "      <td>3.719462</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>AT3G45030</td>\n",
       "      <td>3.347976</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>AT5G58420</td>\n",
       "      <td>3.306624</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>AT1G48630</td>\n",
       "      <td>3.597568</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene    log2FC  p_val  p_val_adj  cluster\n",
       "0    AT2G46890  5.562217    0.0        0.0        1\n",
       "1    AT3G48185  5.308359    0.0        0.0        1\n",
       "2    AT2G40480  4.601549    0.0        0.0        1\n",
       "3    AT1G25530  5.839280    0.0        0.0        1\n",
       "4    AT5G57770  5.368637    0.0        0.0        1\n",
       "..         ...       ...    ...        ...      ...\n",
       "145  AT5G22440  3.521461    0.0        0.0        2\n",
       "146  AT3G58700  3.719462    0.0        0.0        2\n",
       "147  AT3G45030  3.347976    0.0        0.0        2\n",
       "148  AT5G58420  3.306624    0.0        0.0        2\n",
       "149  AT1G48630  3.597568    0.0        0.0        2\n",
       "\n",
       "[150 rows x 5 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enrichment analysis completed. Results saved to: /home/sah2p/ondemand/singlecell_data/Arabidopsis/hypergeometric_enrichment_results.csv\n",
      "    Cluster               CellType  Reference_Size  Cluster_Size  Overlap  \\\n",
      "3         3            Root cortex              10            10        5   \n",
      "7         8  Phloem pole pericycle              10            10        4   \n",
      "5         5               Non-hair              10            10        4   \n",
      "4         4        Root endodermis              10            10        4   \n",
      "1         1        Root procambium              10            10        4   \n",
      "11       14          Sieve element              10            10        3   \n",
      "8         9              Root hair              10            10        3   \n",
      "9        11              Metaxylem              10            10        2   \n",
      "6         7         Companion cell              10            10        2   \n",
      "0         0   Xylem pole pericycle              10            10        2   \n",
      "\n",
      "         p_value         p_adj  \n",
      "3   4.756268e-16  5.707522e-15  \n",
      "7   1.822446e-12  4.373870e-12  \n",
      "5   1.822446e-12  5.467337e-12  \n",
      "4   1.822446e-12  7.289783e-12  \n",
      "1   1.822446e-12  1.093467e-11  \n",
      "11  4.104453e-09  7.036205e-09  \n",
      "8   4.104453e-09  8.208906e-09  \n",
      "9   5.308613e-06  6.370336e-06  \n",
      "6   5.308613e-06  7.078151e-06  \n",
      "0   5.308613e-06  7.962920e-06  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import hypergeom\n",
    "\n",
    "# ==== STEP 1: Load Your Clustered Genes ====\n",
    "# Ensure your file has two columns: \"cluster\" and \"gene\"\n",
    "cluster_df = pd.read_csv(clustered_gene_file)\n",
    "\n",
    "# ==== STEP 2: Load Reference Marker Genes ====\n",
    "# Ensure the file has \"clusterName\" and \"gene\"\n",
    "marker_df = pd.read_csv(marker_gene_file)\n",
    "\n",
    "# ==== STEP 3: Convert Data to Dictionary ====\n",
    "# Convert to dictionary for easy lookup\n",
    "cluster_dict = cluster_df.groupby(\"cluster\")[\"gene\"].apply(set).to_dict()\n",
    "marker_dict = marker_df.groupby(\"clusterName\")[\"gene\"].apply(set).to_dict()\n",
    "\n",
    "# Total genes in the genome (N)\n",
    "N = 27600  # Approximate Arabidopsis gene count\n",
    "\n",
    "# ==== STEP 4: Compute Hypergeometric Enrichment ====\n",
    "enrichment_results = []\n",
    "\n",
    "for cluster, cluster_genes in cluster_dict.items():\n",
    "    for celltype, marker_genes in marker_dict.items():\n",
    "        K = len(marker_genes)  # Genes in the reference cell type\n",
    "        n = len(cluster_genes)  # Genes in your test cluster\n",
    "        k = len(cluster_genes & marker_genes)  # Overlapping genes\n",
    "\n",
    "        if k > 0:  # Only compute if there is some overlap\n",
    "            p_value = hypergeom.sf(k-1, N, K, n)  # Survival function (1 - CDF)\n",
    "            enrichment_results.append([cluster, celltype, K, n, k, p_value])\n",
    "\n",
    "# Convert results to DataFrame\n",
    "enrichment_df = pd.DataFrame(\n",
    "    enrichment_results,\n",
    "    columns=[\"Cluster\", \"CellType\", \"Reference_Size\", \"Cluster_Size\", \"Overlap\", \"p_value\"]\n",
    ")\n",
    "\n",
    "# ==== STEP 5: Adjust p-values (Benjamini-Hochberg FDR) ====\n",
    "enrichment_df[\"p_adj\"] = enrichment_df[\"p_value\"] * len(enrichment_df) / enrichment_df[\"p_value\"].rank(method=\"first\")\n",
    "\n",
    "# Save to CSV\n",
    "output_file = context_path+\"hypergeometric_enrichment_results.csv\"\n",
    "enrichment_df.to_csv(output_file, index=False)\n",
    "\n",
    "print(f\"Enrichment analysis completed. Results saved to: {output_file}\")\n",
    "\n",
    "# ==== STEP 6: Display Top Results (Optional) ====\n",
    "print(enrichment_df.sort_values(\"p_adj\").head(10))  # Show top 10 significant results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orig.ident</th>\n",
       "      <th>Celltype</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Libraries</th>\n",
       "      <th>ACE</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>batch</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>leiden</th>\n",
       "      <th>louvain</th>\n",
       "      <th>leiden_res0_1</th>\n",
       "      <th>leiden_res0_5</th>\n",
       "      <th>leiden_res1</th>\n",
       "      <th>louvain_res0_1</th>\n",
       "      <th>louvain_res0_5</th>\n",
       "      <th>louvain_res1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRX8089019@@_AAACCTGAGAAGGTTT-1-0</th>\n",
       "      <td>SRX8089019</td>\n",
       "      <td>Root procambium</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>1550</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3313.0</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX8089019@@_AAACCTGAGAGTGACC-1-0</th>\n",
       "      <td>SRX8089019</td>\n",
       "      <td>Root cap</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>1715</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>4064.0</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX8089019@@_AAACCTGAGCTAGTCT-1-0</th>\n",
       "      <td>SRX8089019</td>\n",
       "      <td>Root cap</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>1502</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2746.0</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX8089019@@_AAACCTGAGTGGTAGC-1-0</th>\n",
       "      <td>SRX8089019</td>\n",
       "      <td>Root cap</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>939</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2756.0</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX8089019@@_AAACCTGCAATAGCGG-1-0</th>\n",
       "      <td>SRX8089019</td>\n",
       "      <td>Root cap</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>2249</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>5080.0</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX8089021@@_TTTGTCACATGTAAGA-1-2</th>\n",
       "      <td>SRX8089021</td>\n",
       "      <td>Non-hair</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Heat shock</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>987</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2806.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX8089021@@_TTTGTCAGTGCCTGTG-1-2</th>\n",
       "      <td>SRX8089021</td>\n",
       "      <td>Xylem pole pericycle</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Heat shock</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>2665</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>8617.0</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX8089021@@_TTTGTCAGTTGTTTGG-1-2</th>\n",
       "      <td>SRX8089021</td>\n",
       "      <td>Root hair</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Heat shock</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>1013</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2123.0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX8089021@@_TTTGTCATCATGCATG-1-2</th>\n",
       "      <td>SRX8089021</td>\n",
       "      <td>Root hair</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Heat shock</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>1007</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1937.0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX8089021@@_TTTGTCATCATTTGGG-1-2</th>\n",
       "      <td>SRX8089021</td>\n",
       "      <td>Non-hair</td>\n",
       "      <td>SRP166333</td>\n",
       "      <td>Whole root</td>\n",
       "      <td>Root</td>\n",
       "      <td>Heat shock</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>10x Genomics</td>\n",
       "      <td>7 days old seedling</td>\n",
       "      <td>1302</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>4181.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14238 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Orig.ident              Celltype  \\\n",
       "SRX8089019@@_AAACCTGAGAAGGTTT-1-0  SRX8089019       Root procambium   \n",
       "SRX8089019@@_AAACCTGAGAGTGACC-1-0  SRX8089019              Root cap   \n",
       "SRX8089019@@_AAACCTGAGCTAGTCT-1-0  SRX8089019              Root cap   \n",
       "SRX8089019@@_AAACCTGAGTGGTAGC-1-0  SRX8089019              Root cap   \n",
       "SRX8089019@@_AAACCTGCAATAGCGG-1-0  SRX8089019              Root cap   \n",
       "...                                       ...                   ...   \n",
       "SRX8089021@@_TTTGTCACATGTAAGA-1-2  SRX8089021              Non-hair   \n",
       "SRX8089021@@_TTTGTCAGTGCCTGTG-1-2  SRX8089021  Xylem pole pericycle   \n",
       "SRX8089021@@_TTTGTCAGTTGTTTGG-1-2  SRX8089021             Root hair   \n",
       "SRX8089021@@_TTTGTCATCATGCATG-1-2  SRX8089021             Root hair   \n",
       "SRX8089021@@_TTTGTCATCATTTGGG-1-2  SRX8089021              Non-hair   \n",
       "\n",
       "                                     Dataset      Tissue Organ   Condition  \\\n",
       "SRX8089019@@_AAACCTGAGAAGGTTT-1-0  SRP166333  Whole root  Root      Normal   \n",
       "SRX8089019@@_AAACCTGAGAGTGACC-1-0  SRP166333  Whole root  Root      Normal   \n",
       "SRX8089019@@_AAACCTGAGCTAGTCT-1-0  SRP166333  Whole root  Root      Normal   \n",
       "SRX8089019@@_AAACCTGAGTGGTAGC-1-0  SRP166333  Whole root  Root      Normal   \n",
       "SRX8089019@@_AAACCTGCAATAGCGG-1-0  SRP166333  Whole root  Root      Normal   \n",
       "...                                      ...         ...   ...         ...   \n",
       "SRX8089021@@_TTTGTCACATGTAAGA-1-2  SRP166333  Whole root  Root  Heat shock   \n",
       "SRX8089021@@_TTTGTCAGTGCCTGTG-1-2  SRP166333  Whole root  Root  Heat shock   \n",
       "SRX8089021@@_TTTGTCAGTTGTTTGG-1-2  SRP166333  Whole root  Root  Heat shock   \n",
       "SRX8089021@@_TTTGTCATCATGCATG-1-2  SRP166333  Whole root  Root  Heat shock   \n",
       "SRX8089021@@_TTTGTCATCATTTGGG-1-2  SRP166333  Whole root  Root  Heat shock   \n",
       "\n",
       "                                  Genotype     Libraries                  ACE  \\\n",
       "SRX8089019@@_AAACCTGAGAAGGTTT-1-0    Col-0  10x Genomics  7 days old seedling   \n",
       "SRX8089019@@_AAACCTGAGAGTGACC-1-0    Col-0  10x Genomics  7 days old seedling   \n",
       "SRX8089019@@_AAACCTGAGCTAGTCT-1-0    Col-0  10x Genomics  7 days old seedling   \n",
       "SRX8089019@@_AAACCTGAGTGGTAGC-1-0    Col-0  10x Genomics  7 days old seedling   \n",
       "SRX8089019@@_AAACCTGCAATAGCGG-1-0    Col-0  10x Genomics  7 days old seedling   \n",
       "...                                    ...           ...                  ...   \n",
       "SRX8089021@@_TTTGTCACATGTAAGA-1-2    Col-0  10x Genomics  7 days old seedling   \n",
       "SRX8089021@@_TTTGTCAGTGCCTGTG-1-2    Col-0  10x Genomics  7 days old seedling   \n",
       "SRX8089021@@_TTTGTCAGTTGTTTGG-1-2    Col-0  10x Genomics  7 days old seedling   \n",
       "SRX8089021@@_TTTGTCATCATGCATG-1-2    Col-0  10x Genomics  7 days old seedling   \n",
       "SRX8089021@@_TTTGTCATCATTTGGG-1-2    Col-0  10x Genomics  7 days old seedling   \n",
       "\n",
       "                                   n_genes_by_counts  ...  batch  n_counts  \\\n",
       "SRX8089019@@_AAACCTGAGAAGGTTT-1-0               1550  ...      0    3313.0   \n",
       "SRX8089019@@_AAACCTGAGAGTGACC-1-0               1715  ...      0    4064.0   \n",
       "SRX8089019@@_AAACCTGAGCTAGTCT-1-0               1502  ...      0    2746.0   \n",
       "SRX8089019@@_AAACCTGAGTGGTAGC-1-0                939  ...      0    2756.0   \n",
       "SRX8089019@@_AAACCTGCAATAGCGG-1-0               2249  ...      0    5080.0   \n",
       "...                                              ...  ...    ...       ...   \n",
       "SRX8089021@@_TTTGTCACATGTAAGA-1-2                987  ...      2    2806.0   \n",
       "SRX8089021@@_TTTGTCAGTGCCTGTG-1-2               2665  ...      2    8617.0   \n",
       "SRX8089021@@_TTTGTCAGTTGTTTGG-1-2               1013  ...      2    2123.0   \n",
       "SRX8089021@@_TTTGTCATCATGCATG-1-2               1007  ...      2    1937.0   \n",
       "SRX8089021@@_TTTGTCATCATTTGGG-1-2               1302  ...      2    4181.0   \n",
       "\n",
       "                                   leiden  louvain  leiden_res0_1  \\\n",
       "SRX8089019@@_AAACCTGAGAAGGTTT-1-0       9       10              0   \n",
       "SRX8089019@@_AAACCTGAGAGTGACC-1-0       4       12              3   \n",
       "SRX8089019@@_AAACCTGAGCTAGTCT-1-0      10        5              3   \n",
       "SRX8089019@@_AAACCTGAGTGGTAGC-1-0      14       13              3   \n",
       "SRX8089019@@_AAACCTGCAATAGCGG-1-0      10        5              3   \n",
       "...                                   ...      ...            ...   \n",
       "SRX8089021@@_TTTGTCACATGTAAGA-1-2       1        0              2   \n",
       "SRX8089021@@_TTTGTCAGTGCCTGTG-1-2       6        4              1   \n",
       "SRX8089021@@_TTTGTCAGTTGTTTGG-1-2       0        6              5   \n",
       "SRX8089021@@_TTTGTCATCATGCATG-1-2       0        6              5   \n",
       "SRX8089021@@_TTTGTCATCATTTGGG-1-2       1        0              2   \n",
       "\n",
       "                                   leiden_res0_5  leiden_res1  louvain_res0_1  \\\n",
       "SRX8089019@@_AAACCTGAGAAGGTTT-1-0              1            9               0   \n",
       "SRX8089019@@_AAACCTGAGAGTGACC-1-0              6            4               1   \n",
       "SRX8089019@@_AAACCTGAGCTAGTCT-1-0             10           10               1   \n",
       "SRX8089019@@_AAACCTGAGTGGTAGC-1-0             12           14               1   \n",
       "SRX8089019@@_AAACCTGCAATAGCGG-1-0             10           10               1   \n",
       "...                                          ...          ...             ...   \n",
       "SRX8089021@@_TTTGTCACATGTAAGA-1-2              5            1               2   \n",
       "SRX8089021@@_TTTGTCAGTGCCTGTG-1-2              0            5               0   \n",
       "SRX8089021@@_TTTGTCAGTTGTTTGG-1-2              2            0               1   \n",
       "SRX8089021@@_TTTGTCATCATGCATG-1-2              2            0               1   \n",
       "SRX8089021@@_TTTGTCATCATTTGGG-1-2              5            1               2   \n",
       "\n",
       "                                   louvain_res0_5  louvain_res1  \n",
       "SRX8089019@@_AAACCTGAGAAGGTTT-1-0               1            10  \n",
       "SRX8089019@@_AAACCTGAGAGTGACC-1-0               7            12  \n",
       "SRX8089019@@_AAACCTGAGCTAGTCT-1-0               3             5  \n",
       "SRX8089019@@_AAACCTGAGTGGTAGC-1-0               7            13  \n",
       "SRX8089019@@_AAACCTGCAATAGCGG-1-0               3             5  \n",
       "...                                           ...           ...  \n",
       "SRX8089021@@_TTTGTCACATGTAAGA-1-2               4             0  \n",
       "SRX8089021@@_TTTGTCAGTGCCTGTG-1-2               0             4  \n",
       "SRX8089021@@_TTTGTCAGTTGTTTGG-1-2               8             6  \n",
       "SRX8089021@@_TTTGTCATCATGCATG-1-2               8             6  \n",
       "SRX8089021@@_TTTGTCATCATTTGGG-1-2               4             0  \n",
       "\n",
       "[14238 rows x 35 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'predicted_celltype' as categorical\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Annotated AnnData saved to: /home/sah2p/ondemand/singlecell_data/Arabidopsis/hypergeometric_enrichment_results.csv\n"
     ]
    }
   ],
   "source": [
    "# ==== STEP 3: Assign Best-Matching Cell Type ====\n",
    "# Select the best cell type for each cluster (lowest adjusted p-value)\n",
    "best_matches = enrichment_df.sort_values(\"p_adj\").groupby(\"Cluster\").first().reset_index()\n",
    "best_matches = best_matches[[\"Cluster\", \"CellType\", \"p_adj\"]]\n",
    "\n",
    "# Convert cluster column to string (to match AnnData)\n",
    "best_matches[\"Cluster\"] = best_matches[\"Cluster\"].astype(str)\n",
    "\n",
    "# ==== STEP 4: Add Cell Type Annotations to AnnData ====\n",
    "# Ensure cluster labels exist in AnnData\n",
    "if \"leiden_res0_5\" not in adata.obs.columns:\n",
    "    raise ValueError(\"Clustering labels ('leiden') not found in adata.obs. Please run clustering first.\")\n",
    "\n",
    "# Create a mapping from cluster to cell type\n",
    "cluster_to_celltype = dict(zip(best_matches[\"Cluster\"], best_matches[\"CellType\"]))\n",
    "\n",
    "# Assign the cell type annotation\n",
    "adata.obs[\"predicted_celltype\"] = adata.obs[\"leiden_res0_5\"].map(cluster_to_celltype)\n",
    "\n",
    "# ==== STEP 5: Save Updated AnnData Object ====\n",
    "# output_file = \"adata_with_annotations.h5ad\"\n",
    "# adata.write(output_file)\n",
    "\n",
    "# ==== STEP 6: Plot UMAP with Annotated Clusters ====\n",
    "sc.pl.umap(adata, color=[\"leiden_res0_5\", \"predicted_celltype\"], save=dataset+\"umap_predicted_celltypes.png\", frameon=False)\n",
    "\n",
    "print(f\"✅ Annotated AnnData saved to: {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Predicted vs. Ground Truth Comparison Table:\n",
      "Celltype               Companion cell  G2/M phase  Metaxylem  Non-hair  \\\n",
      "predicted_celltype                                                       \n",
      "Companion cell                    603           0         26         4   \n",
      "G2/M phase                         22         104         30        16   \n",
      "Metaxylem                           0           0        496         0   \n",
      "Non-hair                            1           0          2      1164   \n",
      "Phloem pole pericycle               6           0          2         0   \n",
      "Root cap                            0           0          0         1   \n",
      "Root cortex                         2           0          0         6   \n",
      "Root endodermis                     0           0          3         1   \n",
      "Root hair                           0           0          0         7   \n",
      "Root procambium                     2           0          1         5   \n",
      "Sieve element                       3           0          0         0   \n",
      "Xylem pole pericycle                6           0          0         3   \n",
      "\n",
      "Celltype               Phloem pole pericycle  Protoxylem  Root cap  \\\n",
      "predicted_celltype                                                   \n",
      "Companion cell                            34          85         0   \n",
      "G2/M phase                                 1           5       205   \n",
      "Metaxylem                                  0           1         0   \n",
      "Non-hair                                   2           6         0   \n",
      "Phloem pole pericycle                    724           0         0   \n",
      "Root cap                                   0           0       490   \n",
      "Root cortex                                0           2         0   \n",
      "Root endodermis                            0           0         0   \n",
      "Root hair                                  0           1         0   \n",
      "Root procambium                            2           0         3   \n",
      "Sieve element                              0           0         0   \n",
      "Xylem pole pericycle                       7           4         5   \n",
      "\n",
      "Celltype               Root cortex  Root endodermis  Root hair  \\\n",
      "predicted_celltype                                               \n",
      "Companion cell                   1               40          2   \n",
      "G2/M phase                      20               12        488   \n",
      "Metaxylem                        1                0          0   \n",
      "Non-hair                         5                0          5   \n",
      "Phloem pole pericycle            0                1          0   \n",
      "Root cap                         0                0          0   \n",
      "Root cortex                   1278                3          3   \n",
      "Root endodermis                  2             1271          0   \n",
      "Root hair                        1                1        709   \n",
      "Root procambium                  2                2          0   \n",
      "Sieve element                    0                0          0   \n",
      "Xylem pole pericycle             5                0          3   \n",
      "\n",
      "Celltype               Root procambium  S phase  Sieve element  \\\n",
      "predicted_celltype                                               \n",
      "Companion cell                      98       11             17   \n",
      "G2/M phase                          12       61              0   \n",
      "Metaxylem                           11       10              0   \n",
      "Non-hair                             0        0              1   \n",
      "Phloem pole pericycle               13       43              0   \n",
      "Root cap                             0        0              0   \n",
      "Root cortex                          1        0              3   \n",
      "Root endodermis                      2        0              1   \n",
      "Root hair                            1        2              1   \n",
      "Root procambium                   1577       32              2   \n",
      "Sieve element                        1        0            148   \n",
      "Xylem pole pericycle                11       77              0   \n",
      "\n",
      "Celltype               Xylem pole pericycle  \n",
      "predicted_celltype                           \n",
      "Companion cell                            6  \n",
      "G2/M phase                              430  \n",
      "Metaxylem                                 1  \n",
      "Non-hair                                  1  \n",
      "Phloem pole pericycle                    39  \n",
      "Root cap                                  0  \n",
      "Root cortex                               0  \n",
      "Root endodermis                           1  \n",
      "Root hair                                 0  \n",
      "Root procambium                           2  \n",
      "Sieve element                             0  \n",
      "Xylem pole pericycle                   1718  \n",
      "✅ Comparison results saved to: /home/sah2p/ondemand/singlecell_data/Arabidopsis/predicted_vs_ground_truth_comparison.csv\n"
     ]
    }
   ],
   "source": [
    "# # ==== STEP 7: Generate Heatmap for Marker Gene Expression ====\n",
    "# # Select marker genes for visualization\n",
    "# top_marker_genes = best_matches[\"CellType\"].unique()[:15]  # Top 15 representative genes\n",
    "\n",
    "\n",
    "\n",
    "# # ==== STEP 8: Generate Dot Plot ====\n",
    "# sc.pl.dotplot(adata, var_names=top_marker_genes, groupby=\"predicted_celltype\", standard_scale=\"var\")\n",
    "\n",
    "# # ==== STEP 9: Compare Predicted vs. Ground Truth Labels ====\n",
    "comparison_df = adata.obs[[\"predicted_celltype\", \"Celltype\"]]\n",
    "comparison_table = pd.crosstab(comparison_df[\"predicted_celltype\"], comparison_df[\"Celltype\"])\n",
    "\n",
    "# Save the comparison table\n",
    "comparison_output = context_path+\"predicted_vs_ground_truth_comparison.csv\"\n",
    "comparison_table.to_csv(comparison_output)\n",
    "\n",
    "# Display comparison table\n",
    "print(\"✅ Predicted vs. Ground Truth Comparison Table:\")\n",
    "print(comparison_table)\n",
    "\n",
    "print(f\"✅ Comparison results saved to: {comparison_output}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 14238 × 4000\n",
       "    obs: 'Orig.ident', 'Celltype', 'Dataset', 'Tissue', 'Organ', 'Condition', 'Genotype', 'Libraries', 'ACE', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_pt', 'log1p_total_counts_pt', 'pct_counts_pt', 'doublet', 'doublet_score', 'batch', 'n_counts', 'leiden', 'louvain', 'leiden_res0_1', 'leiden_res0_5', 'leiden_res1', 'louvain_res0_1', 'louvain_res0_5', 'louvain_res1', 'predicted_celltype'\n",
       "    var: 'features', 'mt', 'pt', 'n_cells_by_counts-0', 'mean_counts-0', 'log1p_mean_counts-0', 'pct_dropout_by_counts-0', 'total_counts-0', 'log1p_total_counts-0', 'n_cells_by_counts-1', 'mean_counts-1', 'log1p_mean_counts-1', 'pct_dropout_by_counts-1', 'total_counts-1', 'log1p_total_counts-1', 'n_cells_by_counts-2', 'mean_counts-2', 'log1p_mean_counts-2', 'pct_dropout_by_counts-2', 'total_counts-2', 'log1p_total_counts-2', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'\n",
       "    uns: 'Celltype_colors', 'Orig.ident_colors', 'hvg', 'log1p', 'neighbors', 'pca', 'umap', 'leiden', 'louvain', 'leiden_colors', 'louvain_colors', 'leiden_res0_1_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'louvain_res0_1_colors', 'louvain_res0_5_colors', 'louvain_res1_colors', 'rank_genes_groups', 'predicted_celltype_colors'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'logcounts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n"
     ]
    }
   ],
   "source": [
    "sc.pl.umap(adata, color=[\"leiden\", \"Celltype\"], save=dataset+\"umap_celltypes.png\", frameon=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:1251: FutureWarning: The default value of 'ignore' for the `na_action` parameter in pandas.Categorical.map is deprecated and will be changed to 'None' in a future version. Please set na_action to the desired value to avoid seeing this warning\n",
      "  color_vector = pd.Categorical(values.map(color_map))\n",
      "/home/sah2p/ondemand/tools/miniconda3/envs/scanpy/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored\n",
      "  cax = scatter(\n"
     ]
    }
   ],
   "source": [
    "#Umap of predicted vs ground truth\n",
    "sc.pl.umap(adata, color=[\"leiden_res0_5\", \"predicted_celltype\", \"Celltype\"], save=dataset+\"predicted_vs_gt.png\", frameon=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
